### DISCUSSION TITLE
EOFException

### DISCUSSION THREAD

**Discussion by koala775** [rabbitmq/rabbitmq-java-client#811] [General]
* RabbitMQ version：3.6.5
* Erlang version：19.2
* Option：
 When i start java client ,there's no exception ,but when send message trow a exception:
com.rabbitmq.client.AlreadyClosedException:connection is already closed due to connection error;cause:java.io.EOFException.
So,I restart my java client,it's ok.

**Comment by michaelklishin** [comment]
Client libraries cannot avoid I/O exceptions, it's a natural part of distributed system operations.
Java client [has an automatic recovery feature](https://rabbitmq.com/api-guide.html#recovery) that helps in most cases.

See server logs to learn more about what may be going on (or at least what the server observes),
e.g. [connection failures](https://rabbitmq.com/logging.html#logged-events) or [missed heartbeats](https://rabbitmq.com/heartbeats.html) or [proxy side-effects](https://rabbitmq.com/heartbeats.html#tcp-proxies) (if applicable).

FYI, RabbitMQ 3.6 [has reached end of life in May 2018](https://rabbitmq.com/versions.html), so over four years ago. Time to [move on](https://rabbitmq.com/upgrade.html) beyond the One True RabbitMQ Version.

**Reply by koala775** [reply]
Thank for you replaying.I have saw the server's log:
client unexpectedly closed TCP connection.
is reason cause the client closed TCP connection.

**Reply by michaelklishin** [reply]
This is [covered in the connection lifecycle events](https://rabbitmq.com/logging.html#logged-events). RabbitMQ observers that the client abruptly closes TCP connection, while the client observes an EOF (and I/O exception). My guess is that [you have a proxy with lower inactivity timeouts than the heartbeat interval](https://rabbitmq.com/heartbeats.html#tcp-proxies) and a lower (but not too low, see the guide) heartbeat timeout might help.
